File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/WhiteboardRenderer.kt
package com.lextures.android.core.lms

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke

/** Read-only whiteboard canvas renderer (M7.5). */
object WhiteboardRenderer {
    fun drawGrid(scope: DrawScope, isDark: Boolean) {
        val background = if (isDark) Color(0xFF171717) else Color.White
        scope.drawRect(background)
        val dot = if (isDark) Color.White.copy(alpha = 0.18f) else Color.Black.copy(alpha = 0.18f)
        val spacing = 24f
        var x = spacing
        while (x < scope.size.width) {
            var y = spacing
            while (y < scope.size.height) {
                scope.drawCircle(dot, 1f, Offset(x, y))
                y += spacing
            }
            x += spacing
        }
    }

    fun drawElement(scope: DrawScope, element: WhiteboardElement) {
        val color = parseColor(element.color) ?: return
        val stroke = Stroke(width = element.width.toFloat(), cap = androidx.compose.ui.graphics.StrokeCap.Round)
        when (element.type) {
            "stroke" -> {
                val points = element.points()
                if (points.size < 2) return
                val path = Path().apply {
                    moveTo(points[0].x, points[0].y)
                    points.drop(1).forEach { lineTo(it.x, it.y) }
                }
                scope.drawPath(path, color, style = stroke)
            }
            "rect" -> {
                val rect = element.rect() ?: return
                scope.drawRect(color, topLeft = Offset(rect.left, rect.top), size = rect.size, style = stroke)
            }
            "circle" -> {
                val rect = element.ellipse() ?: return
                scope.drawOval(color, topLeft = Offset(rect.left, rect.top), size = rect.size, style = stroke)
            }
            "triangle" -> {
                val points = element.points()
                if (points.size != 3) return
                val path = Path().apply {
                    moveTo(points[0].x, points[0].y)
                    lineTo(points[1].x, points[1].y)
                    lineTo(points[2].x, points[2].y)
                    close()
                }
                scope.drawPath(path, color, style = stroke)
            }
            "line" -> {
                val points = element.points()
                if (points.size != 2) return
                scope.drawLine(color, points[0], points[1], strokeWidth = element.width.toFloat())
            }
        }
    }

    private fun parseColor(raw: String): Color? {
        var hex = raw.trim()
        if (hex.startsWith("#")) hex = hex.drop(1)
        if (hex.length != 6) return null
        val value = hex.toLongOrNull(16) ?: return null
        val red = ((value shr 16) and 0xFF) / 255f
        val green = ((value shr 8) and 0xFF) / 255f
        val blue = (value and 0xFF) / 255f
        return Color(red, green, blue)
    }
}

private fun WhiteboardElement.points(): List<Offset> {
    pts?.let { raw ->
        return raw.mapNotNull { pair ->
            if (pair.size < 2) return@mapNotNull null
            Offset(pair[0].toFloat(), pair[1].toFloat())
        }
    }
    if (x1 != null && y1 != null && x2 != null && y2 != null && x3 != null && y3 != null) {
        return listOf(
            Offset(x1.toFloat(), y1.toFloat()),
            Offset(x2.toFloat(), y2.toFloat()),
            Offset(x3.toFloat(), y3.toFloat()),
        )
    }
    if (x1 != null && y1 != null && x2 != null && y2 != null) {
        return listOf(Offset(x1.toFloat(), y1.toFloat()), Offset(x2.toFloat(), y2.toFloat()))
    }
    return emptyList()
}

private fun WhiteboardElement.rect(): Rect? {
    if (x == null || y == null || w == null || h == null) return null
    return Rect(x.toFloat(), y.toFloat(), (x + w).toFloat(), (y + h).toFloat())
}

private fun WhiteboardElement.ellipse(): Rect? {
    if (cx == null || cy == null || rx == null || ry == null) return null
    return Rect(
        (cx - kotlin.math.abs(rx)).toFloat(),
        (cy - kotlin.math.abs(ry)).toFloat(),
        (cx + kotlin.math.abs(rx)).toFloat(),
        (cy + kotlin.math.abs(ry)).toFloat(),
    )
}

File: clients/ios/Lextures/Core/LMS/WhiteboardRenderer.swift
import SwiftUI

/// Read-only whiteboard canvas renderer (M7.5) — mirrors web `DrawEl` shapes.
enum WhiteboardRenderer {
    static func draw(elements: [WhiteboardElement], in context: GraphicsContext, size: CGSize, isDark: Bool) {
        drawGrid(in: context, size: size, isDark: isDark)
        for element in elements {
            draw(element: element, in: context)
        }
    }

    private static func drawGrid(in context: GraphicsContext, size: CGSize, isDark: Bool) {
        let background = isDark ? Color(red: 0.09, green: 0.09, blue: 0.09) : .white
        context.fill(Path(CGRect(origin: .zero, size: size)), with: .color(background))
        let dotColor = isDark ? Color.white.opacity(0.18) : Color.black.opacity(0.18)
        let spacing: CGFloat = 24
        var gridX = spacing
        while gridX < size.width {
            var gridY = spacing
            while gridY < size.height {
                let rect = CGRect(x: gridX - 1, y: gridY - 1, width: 2, height: 2)
                context.fill(Path(ellipseIn: rect), with: .color(dotColor))
                gridY += spacing
            }
            gridX += spacing
        }
    }

    private static func draw(element: WhiteboardElement, in context: GraphicsContext) {
        guard let color = Color(hex: element.color) else { return }
        var stroke = StrokeStyle(lineWidth: CGFloat(element.width), lineCap: .round, lineJoin: .round)
        switch element.type {
        case "stroke":
            guard element.points.count >= 2 else { return }
            var path = Path()
            path.move(to: element.points[0])
            for point in element.points.dropFirst() {
                path.addLine(to: point)
            }
            context.stroke(path, with: .color(color), style: stroke)
        case "rect":
            guard let rect = element.rect else { return }
            context.stroke(Path(rect), with: .color(color), style: stroke)
        case "circle":
            guard let ellipse = element.ellipse else { return }
            context.stroke(Path(ellipseIn: ellipse), with: .color(color), style: stroke)
        case "triangle":
            guard element.points.count == 3 else { return }
            var path = Path()
            path.move(to: element.points[0])
            path.addLine(to: element.points[1])
            path.addLine(to: element.points[2])
            path.closeSubpath()
            context.stroke(path, with: .color(color), style: stroke)
        case "line":
            guard element.points.count == 2 else { return }
            var path = Path()
            path.move(to: element.points[0])
            path.addLine(to: element.points[1])
            context.stroke(path, with: .color(color), style: stroke)
        default:
            break
        }
    }
}

struct WhiteboardElement: Codable, Hashable {
    var type: String
    var color: String
    var width: Double
    var pts: [[Double]]? = nil
    var rectX: Double? = nil
    var rectY: Double? = nil
    var rectWidth: Double? = nil
    var rectHeight: Double? = nil
    var centerX: Double? = nil
    var centerY: Double? = nil
    var radiusX: Double? = nil
    var radiusY: Double? = nil
    var point1X: Double? = nil
    var point1Y: Double? = nil
    var point2X: Double? = nil
    var point2Y: Double? = nil
    var point3X: Double? = nil
    var point3Y: Double? = nil

    enum CodingKeys: String, CodingKey {
        case type, color, width, pts
        case rectX = "x"
        case rectY = "y"
        case rectWidth = "w"
        case rectHeight = "h"
        case centerX = "cx"
        case centerY = "cy"
        case radiusX = "rx"
        case radiusY = "ry"
        case point1X = "x1"
        case point1Y = "y1"
        case point2X = "x2"
        case point2Y = "y2"
        case point3X = "x3"
        case point3Y = "y3"
    }

    var points: [CGPoint] {
        if let pts {
            return pts.compactMap { pair in
                guard pair.count >= 2 else { return nil }
                return CGPoint(x: pair[0], y: pair[1])
            }
        }
        if let point1X, let point1Y, let point2X, let point2Y, let point3X, let point3Y {
            return [CGPoint(x: point1X, y: point1Y), CGPoint(x: point2X, y: point2Y), CGPoint(x: point3X, y: point3Y)]
        }
        if let point1X, let point1Y, let point2X, let point2Y {
            return [CGPoint(x: point1X, y: point1Y), CGPoint(x: point2X, y: point2Y)]
        }
        return []
    }

    var rect: CGRect? {
        guard let rectX, let rectY, let rectWidth, let rectHeight else { return nil }
        return CGRect(x: rectX, y: rectY, width: rectWidth, height: rectHeight)
    }

    var ellipse: CGRect? {
        guard let centerX, let centerY, let radiusX, let radiusY else { return nil }
        return CGRect(x: centerX - abs(radiusX), y: centerY - abs(radiusY), width: abs(radiusX) * 2, height: abs(radiusY) * 2)
    }
}

private extension Color {
    init?(hex: String) {
        var raw = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        if raw.hasPrefix("#") { raw.removeFirst() }
        guard raw.count == 6, let value = UInt64(raw, radix: 16) else { return nil }
        let red = Double((value >> 16) & 0xFF) / 255
        let green = Double((value >> 8) & 0xFF) / 255
        let blue = Double(value & 0xFF) / 255
        self.init(red: red, green: green, blue: blue)
    }
}
